File: src/levels/L02-factory-method/ts/singleton.ts
class DrinkFactory {
  private static _instance?: DrinkFactory

  // Una sola fábrica para todas las sucursales.
  static get instance() { return (this._instance ??= new DrinkFactory()) }

  // region: DrinkFactory#create
  create() {
    return new Latte() // la instancia única no sabe de qué sucursal viene el pedido
  }
  // endregion
}

class OrderFlow {
  constructor(private barista: Barista) {}
  take(order: Order) { this.barista.prepare(DrinkFactory.instance.create()) }
}

File: src/levels/L02-factory-method/singleton.rb
class DrinkFactory
  # Una sola fábrica para todas las sucursales.
  def self.instance = @instance ||= new

  # region: DrinkFactory#create
  def create
    Latte.new # la instancia única no sabe de qué sucursal viene el pedido
  end
  # endregion
end

class OrderFlow
  def initialize(barista) = @barista = barista
  def take(order) = @barista.prepare(DrinkFactory.instance.create)
end
